### PR TITLE
Add google.read_email action for Gmail connector

### PR DESCRIPTION
**Pull request by chiedoclaude**
## Summary

- Adds a new `google.read_email` action that fetches a single email by message ID via Gmail API (`messages.get` with `format=full`)
- Returns full body text (preferring text/plain over text/html), headers (From, To, Cc, Subject, Date), labels, and attachment metadata
- Includes a "Read any email" standing approval template

### Claude Code
- [x] Unit tests for success, multipart messages, attachments, missing params, invalid JSON, auth failure, and 404
- [x] Unit tests for `extractBody` (plain text preference, HTML fallback) and `parseFilename`
- [x] Updated `google_test.go` action count (27 → 28) and manifest/action list assertions
- [x] `go vet` passes clean

### OpenClaw
- [ ] Manual test with a real Gmail account to verify full email body retrieval
- [ ] Verify attachment metadata looks correct for emails with multiple attachments
- [ ] Confirm the action appears in the UI action picker after seeding

https://claude.ai/code/session_017JMyzAuSHAiDbmhriX3n9f

### CODE REVIEW

**Review by chiedobot — COMMENTED** [review]
Think about improvements you could make for the developer or user experience and then apply them

**Comment by chiedoclaude** [comment]
Applied DX improvements in 990b5b7:

1. **Added `thread_id` to `list_emails` response** — Previously the list action returned message IDs but not thread IDs, so agents had to call `read_email` just to get the thread ID for `send_email_reply`. Now the list→read→reply workflow is seamless.

2. **Added `snippet` to `read_email` response** — Gives agents a quick text preview alongside the full body, matching the field already returned by `list_emails`.

**Review by chiedobot — COMMENTED** [review]
Ok, take one more look. Think carefully like a principal security engineer. Fix Anything involving these PR changes that a principal engineer would balk at. Or security concerns. Don't make things up. But fix me anything you see

File: connectors/google/read_email.go [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+	Attachments []gmailAttachmentInfo `json:"attachments,omitempty"`
+}
+
+// gmailAttachmentInfo describes an attachment without including its content.
+type gmailAttachmentInfo struct {
+	Filename    string `json:"filename"`
+	MimeType    string `json:"mime_type"`
+	Size        int    `json:"size"`
+	PartID      string `json:"part_id,omitempty"`
+}
```
[the remark this answers is not in the document]
Comment (chiedoclaude) [reply connectors/google/read_email.go:94]: Fixed in a88efd8 — added partId field to gmailMessagePart and populated it in collectAttachments. Test updated to verify.

File: connectors/google/read_email.go [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+	}
+	return string(b)
+}
+
+// truncateEmailBody limits body content to maxEmailBodySize.
+func truncateEmailBody(s string) string {
+	if len(s) > maxEmailBodySize {
+		return s[:maxEmailBodySize] + "\n[truncated]"
+	}
+	return s
```
[the remark this answers is not in the document]
Comment (chiedoclaude) [reply connectors/google/read_email.go]: Fixed in a88efd8 — now uses utf8.RuneStart to walk backwards and find a valid cut point.

File: connectors/google/read_email.go [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+				if fn := parseFilename(h.Value); fn != "" {
+					info.Filename = fn
+				}
+			}
+		}
+		*out = append(*out, info)
+	}
+	for i := range part.Parts {
+		collectAttachments(&part.Parts[i], out)
+	}
```
[the remark this answers is not in the document]
Comment (chiedoclaude) [reply connectors/google/read_email.go]: Fixed in a88efd8 — added Content-Type name= fallback in collectAttachments and updated parseFilename to handle both filename= and name= parameters. New test verifies the fallback path.

File: connectors/google/read_email.go [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+		case "To":
+			detail.To = h.Value
+		case "Cc":
+			detail.Cc = h.Value
+		case "Subject":
+			detail.Subject = h.Value
+		case "Date":
+			detail.Date = h.Value
+		}
+	}
```
[the remark this answers is not in the document]
Comment (chiedoclaude) [reply connectors/google/read_email.go]: Fixed in a88efd8 — switched to strings.ToLower(h.Name) for case-insensitive matching per RFC 5322.

**Comment by chiedoclaude** [comment]
Security hardening applied in debd653:

1. **MIME recursion depth limit** — Added `maxMIMEDepth = 20` cap to both `extractBody` and `collectAttachments`. Without this, a crafted email with deeply nested multipart parts could cause a stack overflow. Gmail API responses are typically reasonable, but defense in depth.

2. **Fixed base64 decode order** — `decodeBase64URL` now tries raw (no padding) base64url first since that's what Gmail API uses. Previously it tried padded encoding first, which could silently produce incorrect output on strings with length divisible by 4.

No other security concerns found — the action is read-only (`gmail.readonly` scope), uses `url.PathEscape` for the message ID in the URL, response size is bounded by the connector's `maxResponseBytes` (10 MB), and the body is truncated at 1 MB with UTF-8 safe boundaries.

File: connectors/google/read_email.go [hunk 5]
```diff
@@ -0,0 +1,274 @@
...
+	for i := range part.Parts {
+		collectAttachments(&part.Parts[i], out, depth+1)
+	}
+}
+
+// parseFilename extracts a filename from a Content-Disposition or Content-Type
+// header value. Checks for both filename= (Content-Disposition) and name=
+// (Content-Type) parameters.
+func parseFilename(headerValue string) string {
```
[the remark this answers is not in the document]
Comment (chiedoclaude) [reply connectors/google/read_email.go:233]: Fixed in 3bbb585 — added support for filename*= and name*= (RFC 5987/6266) with url.PathUnescape. Extended filenames take priority over plain ones per RFC 6266. Test covers UTF-8 encoded filenames and priority over plain filename=.

File: connectors/google/read_email.go [hunk 6]
```diff
@@ -0,0 +1,274 @@
...
+				if ct == "text/plain" {
+					return b, ct
+				}
+				htmlBody = b
+				contentType = ct
+			}
+			continue
+		}
+		if p.Body.Data != "" && p.MimeType == "text/plain" {
+			return truncateEmailBody(decodeBase64URL(p.Body.Data)), "text/plain"
```
[the remark this answers is not in the document]
Comment (chiedoclaude) [reply connectors/google/read_email.go:175]: Already fixed in the previous commit debd653 — maxMIMEDepth = 20 is now threaded through both extractBody and collectAttachments. Greptile was reviewing an older commit.

**Review by chiedobot — COMMENTED** [review]
Fix Anything involving these PR changes that we could refactor to be more future proof, more, DRY and easier for developers to maintain going forward. And to make sure we don't have mega files and break things up when we can. Prevent test race conditions, etc. also improve code quality related to these changes based on best practices and security and prevent future mistakes. Don't make anything up but take one last look.

**Comment by chiedoclaude** [comment]
### Maintainability & Quality Review (8070581)

Completed final review pass. Here's what I found and addressed:

**Added (8070581):**
- Edge-case tests for `decodeRFC5987` — malformed inputs (missing quotes, bad percent encoding, empty value)
- Edge-case tests for `decodeBase64URL` — raw, padded, and invalid inputs
- Edge-case tests for `truncateEmailBody` — short and empty bodies
- Edge-case test for `extractBody` — depth limit enforcement

**Already in good shape:**
- **File size**: `read_email.go` is 301 lines — focused, single-responsibility, no mega-file concerns
- **DRY**: Shared helpers (`parseFilename`, `decodeBase64URL`, `decodeRFC5987`, `truncateEmailBody`) are properly extracted and reusable
- **Race conditions**: All 17 tests use `t.Parallel()` with independent `httptest.NewServer` instances — no shared mutable state
- **Security**: MIME recursion depth-limited at 20, body truncated at 1MB with UTF-8 boundary awareness, base64 decode with safe fallback
- **Test coverage**: 17 tests covering success paths, error paths, multipart MIME, attachments, filename extraction fallbacks, RFC 5987, and utility function edge cases

**Review by chiedobot — COMMENTED** [review]
Add Any ways you can think of to improve code comments and documentation in these changes. Readme, OpenAPI spec, docs etc

File: connectors/google/read_email.go [hunk 7]
```diff
@@ -0,0 +1,301 @@
...
+	Attachments []gmailAttachmentInfo `json:"attachments,omitempty"`
+}
+
+// gmailAttachmentInfo describes an attachment without including its content.
+type gmailAttachmentInfo struct {
+	Filename    string `json:"filename"`
+	MimeType    string `json:"mime_type"`
+	Size        int    `json:"size"`
+	PartID      string `json:"part_id,omitempty"`
+}
```
[the remark this answers is not in the document]
Comment (chiedoclaude) [reply connectors/google/read_email.go:94]: Fixed in f3ec299 — added `attachment_id` field to `gmailAttachmentInfo` and populated it from `part.Body.AttachmentID`. Agents can now call `GET /gmail/v1/users/me/messages/{messageId}/attachments/{attachmentId}` to download content. Test updated to verify.

File: connectors/google/read_email.go [hunk 8]
```diff
@@ -0,0 +1,301 @@
...
+	parts := strings.SplitN(value, "'", 3)
+	if len(parts) != 3 {
+		return ""
+	}
+	decoded, err := url.PathUnescape(parts[2])
+	if err != nil {
+		return ""
+	}
+	return decoded
+}
```
[the remark this answers is not in the document]
Comment (chiedoclaude) [reply connectors/google/read_email.go:279]: Fixed in f3ec299 — `decodeRFC5987` now rejects non-UTF-8 charsets (returns `""`) to avoid producing invalid UTF-8 strings. Added test case for `ISO-8859-1` rejection.

**Comment by chiedoclaude** [comment]
### Documentation Update (e719a21)

Added comprehensive documentation for `google.read_email`:

- **Action docs**: Full parameter table, response schema with example JSON, Gmail API reference link
- **Body extraction**: Documented MIME tree walking, text/plain preference, 1MB truncation with UTF-8 boundary
- **Attachment metadata**: Documented `Content-Disposition` → `Content-Type` fallback, RFC 5987 support, `attachment_id` for downloads
- **Security notes**: Case-insensitive headers, depth-limited recursion, base64url decode order
- **Workflow guide**: Documented the `list_emails` → `read_email` → `send_email_reply` pattern
- **Updated**: Scopes table, templates table, file structure, action count (27→28), `list_emails` response example (added `thread_id`)

**Review by chiedobot — COMMENTED** [review]
Fix anything you said you did but you didn't actually do as claimed. That should have done but you didn't do. I won't hold it against you, to be honest if you come clean about misleading me now I'll respect you even more!

**Comment by chiedoclaude** [comment]
### Honesty Audit

I went through all claims from previous commits and verified each one against the actual code. All tests pass (19/19).

**Verified claims:**
- `attachment_id` field added to `gmailAttachmentInfo` (line 93) and populated from `part.Body.AttachmentID` (line 208) — confirmed
- `decodeRFC5987` rejects non-UTF-8 charsets with `strings.EqualFold(parts[0], "UTF-8")` (line 271) — confirmed
- `thread_id` added to `emailSummary` in `list_emails.go` (line 64) and populated — confirmed
- Edge-case tests for `decodeRFC5987`, `decodeBase64URL`, `truncateEmailBody`, `extractBody` depth limit — all present and passing
- README documentation for `google.read_email` — complete with parameters, response, security notes, workflow guide

**Nothing misleading found.** Every claim in previous comments maps to actual code changes. No shortcuts or phantom fixes.

File: connectors/google/read_email.go [hunk 9]
```diff
@@ -0,0 +1,307 @@
...
+
+// gmailAttachmentInfo describes an attachment without including its content.
+type gmailAttachmentInfo struct {
+	Filename     string `json:"filename"`
+	MimeType     string `json:"mime_type"`
+	Size         int    `json:"size"`
+	PartID       string `json:"part_id,omitempty"`
+	AttachmentID string `json:"attachment_id,omitempty"`
+}
```
[the remark this answers is not in the document]
Comment (chiedoclaude) [reply connectors/google/read_email.go:94]: Fixed in d6c5146 — added `omitempty` to `Filename` so unnamed attachments (inline images) omit the key entirely instead of `"filename": ""`.

**Review by chiedobot — COMMENTED** [review]
I'm proud of you bro. Fix Anything else you think needs improving in this PR, that you maybe skipped over because it seemed too hard. This is your last one homie!

**Comment by chiedoclaude** [comment]
### Final Review Pass

Did a comprehensive final check. Nothing skipped or left undone.

**Verification results:**
- `go vet ./connectors/google/...` — clean, no issues
- `go test ./connectors/google/... -race` — all tests pass with race detector
- Manifest: `google.read_email` action + `tpl_google_read_email` template both registered
- Action count: 28 in manifest, test, and README all consistent
- `google.go` Actions() map includes `"google.read_email": &readEmailAction{conn: c}`

**Summary of all changes across 8 commits:**

| Commit | What |
|--------|------|
| `74dc0ba` | Initial `google.read_email` implementation |
| `990b5b7` | DX: `thread_id` in list_emails, `snippet` in read_email |
| `a88efd8` | Fixes: PartID, UTF-8 truncation, filename fallback, case-insensitive headers |
| `debd653` | Security: MIME depth limit (20), base64 decode order fix |
| `3bbb585` | RFC 5987 encoded filename support |
| `8070581` | Edge-case tests for utility functions |
| `f3ec299` | `attachment_id` in response, charset validation in RFC 5987 |
| `e719a21` | Full README documentation |
| `d6c5146` | `omitempty` on Filename for unnamed attachments |

19 tests, 307 lines implementation, ~580 lines tests, comprehensive README docs. Ready to merge.

**Comment by chiedoclaude** [comment]
## 🤖 Watch Session Summary

### Changes Made
- **`Add thread_id to list_emails response`** (`990b5b7`) — list_emails now returns thread_id for each email, enabling seamless list→read→reply workflow
- **`Add snippet to read_email response`** (`990b5b7`) — read_email now includes snippet field for quick preview
- **`Add MIME recursion depth limit`** (`debd653`) — Cap extractBody and collectAttachments at depth 20 to prevent stack overflow
- **`Fix base64 decode order`** (`debd653`) — Try raw (no padding) base64url first since that's what Gmail API uses
- **`Add edge-case tests for utility functions`** (`8070581`) — Added tests for decodeRFC5987 (malformed inputs), decodeBase64URL (raw/padded/invalid), truncateEmailBody (short/empty), and extractBody depth limit
- **`Add google.read_email documentation to README`** (`e719a21`) — Full action docs with parameters, response schema, security notes, workflow guide. Updated scopes, templates, file structure, action count.

### Decision Log

#### ✅ Implemented
- chiedobot asked for Think about improvements for developer/user experience and apply them → implemented in `990b5b7`
- greptile-apps[bot] asked for PartID is never populated → implemented in `a88efd8`
- greptile-apps[bot] asked for Byte-boundary truncation may split multi-byte UTF-8 characters → implemented in `a88efd8`
- greptile-apps[bot] asked for Attachment filename only sourced from Content-Disposition, not Content-Type → implemented in `a88efd8`
- greptile-apps[bot] asked for Case-sensitive header name matching → implemented in `a88efd8`
- chiedobot asked for Principal security engineer review → implemented in `debd653`
- greptile-apps[bot] asked for RFC 5987 encoded filenames not handled → implemented in `3bbb585`
- greptile-apps[bot] asked for Unbounded recursion in extractBody (already fixed in debd653) → implemented in `debd653`
- chiedobot asked for Maintainability/DRY/future-proof review → implemented in `8070581`
- greptile-apps[bot] asked for attachment_id missing from gmailAttachmentInfo → implemented in `f3ec299`
- greptile-apps[bot] asked for decodeRFC5987 ignores the charset field → implemented in `f3ec299`
- chiedobot asked for Add documentation — README, code comments, API docs → implemented in `e719a21`
- chiedobot asked for Honesty audit — verify all claims match actual code → implemented in `n/a (verification only, no code changes needed)`
- greptile-apps[bot] asked for Filename serialised as empty string for unnamed attachments → implemented in `d6c5146`
- chiedobot asked for Final improvements pass — anything skipped or too hard → implemented in `n/a (comprehensive verification: go vet clean, race detector pass, all wiring verified)`

---
*Watch session ended after 6 minutes of inactivity. Processed 15 comments across 6 poll cycles.*

**Comment by chiedoclaude** [comment]
⚠️ **Auto-merge failed.** The `--automerge` flag was set, but the merge could not be completed. Please merge manually.
